### PR TITLE
Remove `Nothing` from the Sampler union type

### PR DESCRIPTION
**Pull request by KDr2**
Try to fix https://github.com/TuringLang/Turing.jl/issues/537

### CODE REVIEW

File: src/core/compiler.jl [hunk 1]
```diff
@@ -26,7 +26,7 @@ end
 
 
 function wrong_dist_errormsg(l)
-    return "Right-hand side of a ~ must be subtype of Distribution or a vector of" * 
+    return "Right-hand side of a ~ must be subtype of Distribution or a vector of " *
```
Comment (yebai) [inline src/core/compiler.jl]: Nice to see this fixed!

File: src/core/ad.jl [hunk 2]
```diff
@@ -30,6 +30,7 @@ getchunksize(::T) where {T <: ForwardDiffAD} = getchunksize(T)
 getchunksize(::Type{ForwardDiffAD{chunk}}) where chunk = chunk
 getchunksize(::T) where {T <: Sampler} = getchunksize(T)
 getchunksize(::Type{<:Sampler{T}}) where {T} = getchunksize(T)
+getchunksize(::SampleFromPrior) = getchunksize(Nothing)
```
Comment (yebai) [inline src/core/ad.jl]: Remove `getchunksize(::Nothing) = getchunksize(Nothing)`?

File: src/inference/Inference.jl [hunk 3]
```diff
@@ -185,10 +172,16 @@ function assume(spl::A,
 
 end
 
+
+observe(::Nothing,
```
Comment (yebai) [inline src/inference/Inference.jl]: Nice!

File: src/core/VarReplay.jl [hunk 4]
```diff
@@ -102,14 +103,15 @@ mutable struct VarInfo
     end
 end
 
-function Turing.runmodel!(model::Model, vi::VarInfo, spl::Union{Nothing,Sampler})
+function Turing.runmodel!(model::Model, vi::VarInfo, spl::AbstractSampler)
```
Comment (trappmartin) [inline src/core/VarReplay.jl]: Would it make sense to use a generated function here?

File: src/core/VarReplay.jl [hunk 5]
```diff
@@ -294,7 +296,8 @@ end
 
 # Get all indices of variables belonging to gid or 0
 getidcs(vi::VarInfo) = getidcs(vi, nothing)
-getidcs(vi::VarInfo, spl::Nothing) = filter(i -> vi.gids[i] == 0, 1:length(vi.gids))
+getidcs(vi::VarInfo, ::SampleFromPrior) = filter(i -> vi.gids[i] == 0, 1:length(vi.gids))
+# getidcs(vi::VarInfo, spl::Nothing) = filter(i -> vi.gids[i] == 0, 1:length(vi.gids))
```
Comment (trappmartin) [inline src/core/VarReplay.jl]: Remove comment?

File: src/core/container.jl [hunk 6]
```diff
@@ -1,8 +1,8 @@
 mutable struct Trace
   task  ::  Task
   vi    ::  VarInfo
-  spl   ::  Union{Nothing, Sampler}
-  Trace() = (res = new(); res.vi = VarInfo(); res.spl = nothing; res)
+  spl   ::  AbstractSampler
```
Comment (trappmartin) [inline src/core/container.jl]: Maybe we can make `Trace` a parametric type?

File: src/core/compiler.jl (hunk 1 above)
Comment (KDr2) [reply src/core/compiler.jl]: 😄

File: src/core/ad.jl (hunk 2 above)
Comment (KDr2) [reply src/core/ad.jl]: Done.

File: src/core/VarReplay.jl (hunk 4 above)
Comment (KDr2) [reply src/core/VarReplay.jl]: Hmm... I think it's not a big impact: a generated function only change the `isa` type-check to a multi-dispatch. Anyway, I changed it to a generated function.

File: src/core/VarReplay.jl (hunk 5 above)
Comment (KDr2) [reply src/core/VarReplay.jl]: Done.

File: src/core/container.jl (hunk 6 above)
Comment (KDr2) [reply src/core/container.jl]: Done.

**Pull request merged**